ARROW-8628: [Dev] Wrap docker-compose commands with archery

kszucs opened the pull request:
- extend docker-compose.yml with a custom field to define image hierarchy
- added docker and docker-compose commands to archery
- validate docker-compose configuration's custom fields as well exercising docker-compose's builtin validation
- covered the added features with unittests
- archery command to execute the pull/build/run docker-compose sequences
- archery support for python 3.5, also test it on github actions
- updated all of the github workflows to use the new archery tooling
- updated all of the crossbow tasks to use archery as well
- removed the old Makefile.docker
- fixed archery's package and module definitions, because a non-editable installation caused import errors
- added documentation

wesm commented:
> it's pretty straightforward to use docker-compose to reproduce these builds locally

I see both sides of this -- with some builds requiring 3 or more `docker-compose build` invocations and relying on various environment variables / parameters, there's a lot of room for frustration and user error (speaking from experience). I'm also pretty anti-copypasta in build configurations -- having all the build logic tightly coupled to GHA seems to go against some of the spirit of our prior decoupling from Travis CI, Appveyor. 

Of course, everything has to be well-documented and able to be serviced successfully by most Arrow maintainers

kszucs commented:
@github-actions crossbow submit -g test

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-conda-python-3.7-pandas-latest test-debian-10-cpp test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-debian-10-cpp test-debian-10-go-1.12 test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-debian-10-cpp test-debian-10-go-1.12 test-conda-python-3.7

kszucs commented:
@github-actions crossbow submit test-debian-10-cpp

kszucs commented:
@github-actions crossbow submit test-debian-10-cpp

kszucs commented:
@github-actions crossbow submit -g test

kszucs commented:
@github-actions crossbow submit -g test

fsaintjacques commented on .github/workflows/archery.yml:
```diff
@@ -51,10 +53,12 @@ jobs:
       - name: Install
         working-directory: dev/archery
-        run: pip install pytest responses ruamel.yaml toolz jinja2 -e .
+        run: pip install pytest python-dotenv responses ruamel.yaml toolz jinja2 -e .
```
Why not use pip install -e to ensure that requirements are up to date?

kszucs replied:
There is `-e .` at the and of the command. The additional dependencies listed here are required for testing and crossbow.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+-----
+
+There are multiple ways to execute the docker based builds. The recommented is
+to use the archery tool:
```
Can this link to the main Archery docs? (*Are there Archery docs?*)

nealrichardson replied:
Among the questions I'd have for the archery docs specifically right here:

* What is archery?
* What system requirements does it have? (like, python version?)
* What do I do if the simple instructions don't work?

kszucs replied:
There is a tiny intro under the benchmarking guide. I'm deferring it to a follow-up https://issues.apache.org/jira/browse/ARROW-8688

nealrichardson replied:
An immediate followup, right? I'm not sure that I agree with deferring but I'll allow it as long as it's the next thing you do. It looks like you already added some of the relevant information below, not sure how much more you need.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+Usage
+-----
+
+There are multiple ways to execute the docker based builds. The recommented is
```
```suggestion
There are multiple ways to execute the docker based builds. The recommended is
```

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+
+.. code::bash
+
+    docker-compose pull --ignore-pull-failures conda-cpp
```
Can I run these docker-compose commands myself?

kszucs replied:
Yes.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+
+.. code::bash
+
+    archery docker run conda-python
```
Would I ever call `archery docker` and not immediately put `run` next? If not, why not just `archery run conda-python`?

kszucs replied:
I kept that open for features like listing available build(er)s. I can change this to something like `archery build conda-python` if you find it more convenient.

nealrichardson replied:
Sounds like YAGNI to me, anyone else have an opinion?

kszucs replied:
Just added the `archery docker images` command. We can refine the CLI API later and even keep the old commands if we choose to.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+    docker-compose build conda-python
+    docker-compose run --rm conda-python
+
+To disable the image pulling:
```
Is there `archery docker --help` or something that tell me all of the flags like this?

kszucs replied:
Yes.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+    docker-compose build --no-cache conda-python
+    docker-compose run --rm conda-python
+
+To disable the cache only for the leaf image - useful to force building the
```
What's a leaf image?

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+
+    PANDAS=master archery docker run --no-cache-leaf conda-python
+
+``PANDAS`` is a build time parameter, see the defaults in the .env file.
```
This needs more explanation, particularly to distinguish build-time parameters from environment variables

kszucs replied:
Described below, added a reference.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+    archery docker run --no-build conda-python
+
+In order to alter the runtime parameters pass environment variables to the
+container in execution time:
```
That's what `-e` means?

kszucs replied:
Shorthand for `--env`. Using the full parameter name.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+
+.. code::bash
+
+    archery docker run -e CMAKE_BUILD_TYPE=release ubuntu-cpp
```
For C++, can I pass all cmake flags that way?

kszucs replied:
The ones defined in cpp_build.sh. Added a reference for it.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+
+    archery docker run -e CMAKE_BUILD_TYPE=release ubuntu-cpp
+
+Starting an interactive bash session for debugging:
```
... by doing what? (Please be explicit so I don't have to diff the examples to figure out what to do)

kszucs replied:
I'm not sure what do you mean, updated to be a bit more verbose.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+This helps reducing duplications and preventing a series of maintenance, but
+makes the docker-compose configuration more complicated.
+
+Docker Build Parameters
```
Ok, these are the build parameters. Can you link to this section where you mention that above?

kszucs replied:
Added.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+~~~~~~~~~~~~~
+
+The scripts maintainted under ci/scripts directory should be kept as minimal as
+possible to be responsible only a subset of tasks.
```
```suggestion
possible to be responsible for only a subset of tasks.
```

What subset? Why minimal?

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+possible to be responsible only a subset of tasks.
+
+The parametrization is done through environment variables with sane defaults to
+keep the build configurations declerative. Note that these parameters are
```
```suggestion
keep the build configurations declarative. Note that these parameters are
```

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+The scripts maintainted under ci/scripts directory should be kept as minimal as
+possible to be responsible only a subset of tasks.
+
+The parametrization is done through environment variables with sane defaults to
```
```suggestion
The parametrization is done through environment variables with useful defaults to
```

What parametrization?

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+The parametrization is done through environment variables with sane defaults to
+keep the build configurations declerative. Note that these parameters are
+different from the ones described previously, these are affecting the runtime
+behaviour of the builds scripts within the containers.
```
```suggestion
behavior of the builds scripts within the containers.
```

Also this sentence is not clear to me.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+
+A good example is cpp_build.sh build script which forwards environment
+variables as CMake options - so the same scripts can be invoked in various
+configurations without the necessity of chaning it. For examples see how the
```
chaning?

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,143 @@
+Adding New Images
+~~~~~~~~~~~~~~~~~
+
+See more in the docker-compose.yml
```
I don't understand

nealrichardson commented on docs/source/developers/integration.rst:
```diff
@@ -62,9 +62,9 @@ simplify this, we provide a Makefile.
 
-   # Using the makefile with proper image dependency resolution
-   make -f Makefile.docker python
+   # Using the archery with proper image dependency resolution
```
```suggestion
   # Using archery with proper image dependency resolution
```

fsaintjacques commented on dev/archery/archery/docker.py:
```diff
@@ -0,0 +1,152 @@
+
+    def push(self, image, user, password):
+        try:
+            Docker().run('login', '-u', user, '-p', password)
```
Use `--password-stdin`, otherwise they'll leak in the process argv via ps/top/etc and you won't need the next exception re-raising.

kszucs replied:
Created follow-up https://issues.apache.org/jira/browse/ARROW-8698

fsaintjacques commented on dev/archery/archery/docker.py:
```diff
@@ -0,0 +1,152 @@
+        # build all parents
+        for parent in self.nodes[image]:
+            if cache:
+                self._execute('pull', '--ignore-pull-failures', parent,
```
Refactor into a small local function
```python
def build_target(target, cond, env):
  if cond:
    self._execute('pull', '--ignore-pull-failures', target, env=env)
    self._execute('build', target, env=env)
  else:
    self._execute('build', '--no-cache', target, env=env)

```

fsaintjacques commented on dev/archery/archery/cli.py:
```diff
@@ -643,5 +647,78 @@ def trigger_bot(event_name, event_payload, arrow_token, crossbow_token):
 
 
+@archery.group('docker')
+@click.option('--config', '-c', type=click.Path(exists=True),
```
Use `--src` like the other commands. See [archery/utils/source.py,](https://github.com/apache/arrow/blob/master/dev/archery/archery/utils/source.py) you may want to add a `docker_compose` property.

kszucs commented:
Thanks for the reviews, addressing them.

kszucs commented on .github/workflows/java.yml:
```diff
@@ -38,6 +38,8 @@ on:
   DOCKER_BUILDKIT: 0
   COMPOSE_DOCKER_CLI_BUILD: 1
+  ARCHERY_DOCKER_USER: ${{ secrets.DOCKERHUB_TOKEN }}
+  ARCHERY_DOCKER_PASSWORD: ${{ secrets.DOCKERHUB_USER }}
```
Missed that. I'm not sure what would the best way to reduce the maintenance cost of this yml jungle. One possibility to use a crossbow like yml generator, but it has numerous disadvantages.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+    archery docker --help
+    archery docker run --help
+
+``archery docker run`` tries to provide a similar interface to
```
Does it though? `run` here seems to entail `build && run`.

kszucs replied:
The command's interface tries to correspond to `docker-compose **run**` with the addition to build the required images. 
```bash

docker-compose run <image> <cmd> 
archery docker run <image> <cmd>

docker-compose run --env VAR_A=A -e VAR_B=B <image> <cmd> 
archery docker run --env VAR_A=A -e VAR_B=B <image> <cmd>
```

kszucs replied:
Removed it.

nealrichardson commented on docs/source/example1.dat:
```diff
@@ -0,0 +1 @@
+some data
```
Was this supposed to be checked in?

kszucs replied:
Nope.

nealrichardson commented on docs/source/developers/integration.rst:
```diff
@@ -56,15 +56,19 @@ build mount is used for caching and sharing state between staged images.
+
+   archery docker run conda-python
+
+See the `Running Docker Builds`_ for more details.
```
```suggestion
See `Running Docker Builds`_ for more details.
```

kszucs replied:
Updated.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+    docker-compose build conda-python
+    docker-compose build --no-cache conda-python-pandas
+    docker-compose run --rm conda-python-pandas
+
```
Wish: a `--dry-run` flag that just prints the commands that would be run. Would be helpful for those who are more familiar with `docker-compose` and for those who need to edit docker-compose.yml

kszucs replied:
Added.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+Installation
+~~~~~~~~~~~~
+
+Requires ``python>=3.5``. It is recommended to install archery in ``editable``
```
```suggestion
``archery`` requires ``python>=3.5``. It is recommended to install archery in ``editable``
```

kszucs replied:
Updated.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+**To disable the cache only for the leaf image:**
+
+Useful to force building the development version of a dependency.
+The leaf image is ``conda-python-pandas`` in the example.
```
This still doesn't tell me what a leaf image is.

kszucs replied:
Tried to clarify it.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+    docker-compose build --no-cache conda-python-pandas
+    docker-compose run --rm conda-python-pandas
+
+Note that it doens't pull the conda-python-pandas image and disable the cache
```
```suggestion
Note that it doesn't pull the conda-python-pandas image and disable the cache
```

kszucs replied:
Updated

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+``PANDAS`` is a `build parameter <Docker Build Parameters>`_, see the
+defaults in the .env file.
+
+**To entirely skip building the image:**
```
why would I do this?

kszucs replied:
Added explanation.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+    archery docker run --no-build conda-python
+
+In order to alter the runtime parameters pass docker environment variables to
+the container during its execution:
```
How? (be explicit, don't make me guess by inspecting the example)

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+
+    archery docker run --env CMAKE_BUILD_TYPE=release ubuntu-cpp
+
+See the available C++ in the ``ci/scripts/cpp_build.sh`` script.
```
available C++... what? build flags?

kszucs replied:
Updated.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+
+**Run the image with custom command:**
+
+Custom docker commands may be passed as the second argument. The following
```
second argument? Looks like the fourth to me.

kszucs replied:
Of the `archery docker run` command.

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,197 @@
+are dependent on the C++ implementation, so instead of redefining the
+C++ environment multiple Dockerfiles, we can reuse the exact same base C++
+image when building Glib, Ruby, R and Python bindings.
+This helps reducing duplications and preventing a series of maintenance, but
```
```suggestion
This reduces duplication and streamlines maintenance, but
```

kszucs replied:
Updated.

kszucs commented:
> Some UX improvement I'd like to see:
> 
> * I often fail to run `docker-compose` from the root of the sources leading to cryptic errors because it can't find the `.env`. Could you automatically fix this with a setcwd in archery? This is a nice UX feature for user who don't know this detail.

`archery docker run` now works from any subdirectory of arrow, even when it is installed in non-editable mode. The script handles the reading of dotenv defaults for better integration, so we always override those values.

> * It would be nice if upon failure, we'd get the list of environment variables changes affecting the docker build, e.g. PANDAS=master... printed somewhere such that we can easily reproduce. 

Added.

> * Add a `archery docker list` command so we don't need to look at docker-compose.yml to get the list of available targets. Bonus point if you add a short description

Added and the bonus is deferred to https://issues.apache.org/jira/browse/ARROW-8697

kszucs commented:
@github-actions crossbow submit -g test

nealrichardson commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,224 @@
+eventually causing cache misses full image rebuilds.
+
+If the image has been already built but the cache doesn't work properly, it can
+be useful to skip the build phases:
```
This sounds backwards to me but probably just needs a little explanation. If caching isn't working properly, doesn't that mean I *would* need to build, rather than? Because it can't take it from the cache. I would think that `--no-build` means "use the cache".

kszucs replied:
Tried to illustrate with an example.

pitrou commented on .github/workflows/archery.yml:
```diff
@@ -48,13 +50,15 @@ jobs:
         with:
-          python-version: '3.7'
-      - name: Install
+          python-version: '3.5'
```
Are you sure? That version is starting to get old...

kszucs replied:
On older linux distros this is the default, so I was trying to make it more convenient for non-python developers. We can of course increase the python requirement later.

pitrou commented on .github/workflows/cpp_cron.yml:
```diff
@@ -36,170 +36,69 @@ env:
+            ubuntu: 18.04
     env:
-      DEBIAN: ${{ matrix.debian }}
+      # the defaults here should correspond to the values in .env
```
Why are the defaults required, since each matrix entry lists the required value explicitly?

kszucs replied:
The `docker-compose.yml` doesn't contain default values for these environment variables, they are grabbed from the dotenv file. 

If we don't provide a default value for the env variable in the GHA yml then it'll set it to empty. Since we need to explicitly list the docker volumes in the docker-compose.yml:

```
  amd64-ubuntu-14.04-cache:
  amd64-ubuntu-16.04-cache:
  amd64-ubuntu-18.04-cache:
  amd64-ubuntu-20.04-cache:
```

So an empty UBUNTU variable would look for `amd64-ubuntu--cache` which is not defined, so docker-compose bails out.

In my upcoming GHA cache PR I change the type of the docker volumes which may let us to remove the explicit volume definitions, so perhaps we can remove the default env values from the workflow ymls as well.

pitrou commented on .github/workflows/python.yml:
```diff
@@ -34,103 +34,75 @@ on:
+        - name: conda-python-3.6-pandas-0.23
+          image: conda-python-pandas
+          title: AMD64 Conda Python 3.8 Pandas 0.23
+          python: 3.6
```
Title says "3.8"

kszucs replied:
Nice catch.

pitrou commented on .github/workflows/python.yml:
```diff
@@ -34,103 +34,75 @@ on:
+        - name: conda-python-3.6-pandas-latest
+          image: conda-python-pandas
+          title: AMD64 Conda Python 3.8 Pandas latest
+          python: 3.6
```
Title says "3.8"

pitrou commented on dev/archery/archery/cli.py:
```diff
@@ -650,11 +651,130 @@ def integration(with_all=False, random_seed=12345, **args):
+@click.option('--cache/--no-cache', default=True,
+              help='Try to pull the image and its parents')
+@click.option('--cache-leaf/--no-cache-leaf', default=True,
+              help='Force build the image and its parents')
```
I don't understand why `--cache-leaf` means "force build". IMHO, either the option name or its description is off.

kszucs replied:
I'm open to any other naming. `docker build` and `docker-compose build` use the `--no-cache` option to force rebuild the image, so I tried to stick with it.

pitrou replied:
I mean the description doesn't correspond to the option.

kszucs replied:
Right, updating.

pitrou commented on dev/archery/archery/cli.py:
```diff
@@ -650,11 +651,130 @@ def integration(with_all=False, random_seed=12345, **args):
+@docker_compose.command('images')
+@click.pass_obj
+def docker_compose_images(obj):
+    """Push the generated docker-compose image."""
```
This docstring seems off.

kszucs replied:
Updating.

pitrou commented on dev/archery/archery/cli.py:
```diff
@@ -650,11 +651,130 @@ def integration(with_all=False, random_seed=12345, **args):
+              help='Docker login user')
+@click.option('--password', '-p', required=True,
+              envvar='ARCHERY_DOCKER_PASSWORD',
+              help='Docker login user')
```
This description seems off.

kszucs replied:
Updating.

pitrou commented on dev/archery/archery/testing.py:
```diff
@@ -0,0 +1,63 @@
+class PartialEnv(dict):
+
+    def __eq__(self, other):
+        return self.items() <= other.items()
```
Why do you need such a weird `__eq__` method?

kszucs replied:
I use it to check only the relevant subset of environment variables in the mocked `subprocess.run(..., env=env)` call.

pitrou commented on docs/source/developers/docker.rst:
```diff
@@ -0,0 +1,224 @@
+.. raw:: html
+
+   <!--
```
Can't you copy/paste the copyright block from another .rst file instead? Using raw HTML for this is weird.

kszucs replied:
Updating.

kszucs commented on dev/archery/archery/docker.py:
```diff
@@ -0,0 +1,186 @@
+        else:
+            self.params = {k: v for k, v in params.items() if k in self.dotenv}
+
+        # forward the process' environment variables
```
Try `UBUNTU=20.04 archery docker run ubuntu-cpp`. The `conda-cpp` image always depend on ubuntu 18.04 so the `UBUNTU` env variable is ignored.

Raising a warning about it would need to parse out the arguments used in that particular docker-compose service.

kszucs commented on dev/archery/archery/cli.py:
```diff
@@ -30,14 +31,16 @@
 logging.basicConfig(level=logging.INFO)
 
 
+DEFAULT_ARROW_PATH = Path(__file__).parents[3].absolute()
```
Nope, removing.

kszucs commented:
Hopefully we won't have any regressions, if we do I can handle those in follow-up PRs.

Thanks for the reviews! 

Merging.
